File: sample-app-java/app/controllers/WvypStream.java
package controllers;

import com.ybrikman.ping.javaapi.bigpipe.HtmlStreamHelper;
import com.ybrikman.ping.javaapi.bigpipe.Pagelet;
import com.ybrikman.ping.scalaapi.bigpipe.HtmlStream;
import data.ServiceClient;
import play.libs.F;
import play.mvc.Controller;
import play.mvc.Result;
import play.twirl.api.Html;

import javax.inject.Inject;

public class WvypStream extends Controller {
  private final ServiceClient serviceClient;

  @Inject
  public WvypStream(ServiceClient serviceClient) {
    this.serviceClient = serviceClient;
  }

  public Result index() {
    F.Promise<String> wvypPromise = serviceClient.makeServiceCall("wvyp");
    F.Promise<String> searchPromise = serviceClient.makeServiceCall("search");

    F.Promise<Html> wvypHtmlPromise = wvypPromise.map(str -> views.html.wvyp.wvypCount.apply(Integer.parseInt(str)));
    F.Promise<Html> searchHtmlPromise = searchPromise.map(str -> views.html.wvyp.searchCount.apply(Integer.parseInt(str)));

    Pagelet wvypPagelet = Pagelet.fromHtmlPromise(wvypHtmlPromise, "wvypCount");
    Pagelet searchPagelet = Pagelet.fromHtmlPromise(searchHtmlPromise, "searchCount");

    HtmlStream body = HtmlStreamHelper.fromInterleavedPagelets(wvypPagelet, searchPagelet);

    return ok(HtmlStreamHelper.toChunks(views.stream.wvyp.wvyp.apply(body)));
  }
}

File: sample-app-scala/app/controllers/WvypStream.scala
package controllers

import play.api.mvc.{Action, Controller}
import data.ServiceClient
import play.api.libs.concurrent.Execution.Implicits._
import com.ybrikman.ping.scalaapi.bigpipe.{Pagelet, HtmlStream}
import com.ybrikman.ping.scalaapi.bigpipe.HtmlStreamImplicits._

/**
 * An example standalone module that represents the "Who's Viewed Your Profile" module from the LinkedIn homepage.
 * It's identical to the Wvyp controller, except that the data is streamed to the browser as soon as it's available in
 * small "pagelets", similar to Facebook's BigPipe.
 */
class WvypStream(serviceClient: ServiceClient) extends Controller {

  def index = Action { implicit request =>
    val wvypCountFuture = serviceClient.makeServiceCall("wvyp")
    val searchCountFuture = serviceClient.makeServiceCall("search")

    val wvypPagelet = Pagelet.fromHtmlFuture(wvypCountFuture.map(str => views.html.wvyp.wvypCount(str.toInt)), "wvypCount")
    val searchPagelet = Pagelet.fromHtmlFuture(searchCountFuture.map(str => views.html.wvyp.searchCount(str.toInt)), "searchCount")

    val body = HtmlStream.fromInterleavedPagelets(wvypPagelet, searchPagelet)

    Ok.chunked(views.stream.wvyp.wvyp(body))
  }
}
